Add new Govee light integration

LaggAt opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
I made a missing integration for govee lights and led strips, using their official cloud api.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
# nothing needed, uses config flows. just add the new "govee" integration
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: home-assistant/home-assistant.io#14593

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [x] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-asc+-review%3Aapproved

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

LaggAt committed:
added integration for Govee light strips, basic

LaggAt committed:
added integration for Govee light strips, basic

LaggAt committed:
Merge branch 'feature/govee-led-strips' of https://github.com/LaggAt/home-assistant-core into feature/govee-led-strips

LaggAt committed:
added brightness, color, color_temp + fixes

LaggAt committed:
added integration for Govee light strips, basic

LaggAt committed:
added integration for Govee light strips, basic

LaggAt committed:
added brightness, color, color_temp + fixes

LaggAt committed:
added brightness, color, color_temp + fixes

LaggAt committed:
brightness scale fix

LaggAt committed:
brightness scale fix

LaggAt committed:
Merge branch 'feature/govee-led-strips' of https://github.com/LaggAt/home-assistant-core into feature/govee-led-strips

LaggAt committed:
reloadable component, state mgmt, more devices

LaggAt committed:
reloadable component, state mgmt, more devices

LaggAt committed:
use fixed library version for brightness

LaggAt committed:
test config flow, lib version for 429 tracking

LaggAt committed:
test config flow, lib version for 429 tracking

LaggAt committed:
using latest lib for api limit debugging

LaggAt committed:
replace deprecated _LOGGER.warn()

LaggAt committed:
replace deprecated _LOGGER.warn()

LaggAt committed:
use latest lib for better debugging

LaggAt committed:
use latest lib for better debugging

LaggAt committed:
use latest lib for better debugging

LaggAt committed:
Merge branch 'feature/govee-led-strips' of https://github.com/LaggAt/home-assistant-core into feature/govee-led-strips

LaggAt committed:
added global requirements_all.txt and .coveragerc

LaggAt committed:
added global requirements_all.txt and .coveragerc

homeassistant commented:
Hi @LaggAt, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#40476).

Once you do that we will be able to review and accept this pull request.

Thanks!

frenck commented:
⚠️ I was unable to find a documentation PR (matching this code change) in our [documentation repository](https://github.com/home-assistant/home-assistant.io).

Please, update the documentation and open a PR for it (be sure to create a documentation PR against the `next` branch) and link 🔗 this PR and the new documentation PR by mentioning the link to the PR's in both openings posts/descriptions.

🏷 I am adding the `docs-missing` label until this has been resolved.

LaggAt commented:
> ⚠️ I was unable to find a documentation PR (matching this code change) in our [documentation repository](https://github.com/home-assistant/home-assistant.io).
> 
> Please, update the documentation and open a PR for it (be sure to create a documentation PR against the `next` branch) and link 🔗 this PR and the new documentation PR by mentioning the link to the PR's in both openings posts/descriptions.
> 
> 🏷 I am adding the `docs-missing` label until this has been resolved.

will do. Didn't know i was expected to do this before. Excuse me.
I also saw the tests on Py3.7 failing, will fix these on weekend.
Thanks.

LaggAt commented:
Documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/14593

LaggAt committed:
Added python 3.7 tests and support to lib

LaggAt committed:
Added python 3.7 tests and support to lib

LaggAt committed:
updated requirements files

LaggAt committed:
updated requirements files

LaggAt committed:
applying isort

LaggAt committed:
applying isort

LaggAt committed:
isort tests

LaggAt committed:
isort tests

LaggAt committed:
flake8 issues

LaggAt committed:
flake8 issues

LaggAt committed:
formatting code

LaggAt committed:
Merge branch 'feature/govee-led-strips' of https://github.com/LaggAt/home-assistant-core into feature/govee-led-strips

LaggAt committed:
LEARNING_FEEDBACK

LaggAt committed:
Merge remote-tracking branch 'upstream/dev' into feature/govee-led-strips

LaggAt committed:
fix flake8 empty line

LaggAt committed:
flake 8 blank line

LaggAt commented:
A rebase failed miltiple times, therefore i did a merge from dev. Sorry about that.

LaggAt committed:
pylint fixes

LaggAt committed:
Merge commit 'e30acfbfee5d3b7e873b28b98f12559a7c38e8f5' into feature/govee-led-strips

LaggAt committed:
pylint fixes

LaggAt committed:
centralized update using DataUpdateCoordinator

LaggAt committed:
Merge commit 'f09f7feb940cebf82920e19bf1e2ab431947745a' into feature/govee-led-strips

LaggAt committed:
updated requirements + isort

LaggAt committed:
Merge commit 'a8b68dc4f94e32036211ff69cd5c6ab21279af35' into feature/govee-led-strips

LaggAt commented:
I think I'm finally done, may I ask you to check it @frenck ?

dlashua commented:
I packaged this as a custom_component and it's working well for me so far.

LaggAt commented:
> I packaged this as a custom_component and it's working well for me so far.

yes, it's working well for some users too, as you see in this thread:
https://community.home-assistant.io/t/govee-led-strips-integration/228516

Some lights use different brightness levels (0-254, 0-100) in the govee api. I started a new feature branch to address this, but this shouldn't hinder inclusion in the core. This will be a new PR for this.

Could someone remove the **docs-missing** after checking the docs PR? 
https://github.com/home-assistant/home-assistant.io/pull/14593
Thanks.

CoMPaTech commented:
Just as a quick glance, the coverage of `config_flow` is required to have 100%. If I look at the 'files changed' tab it indicates `Added line #L66 was not covered by tests`. 

Additionally: You should only submit `strings.json` and not the translations. You have some commented code in your PR, so that probably should be removed. The empty lists/dicts from manifest can just be discarded (they will be empty by default).

As for docs-missing, not sure if you close (wait a while) and reopen the PR if the bot will pick it up accordingly (I know that works for 'needs-rebase' in docs).

LaggAt commented:
Thanks a lot, I'll close the PR for now to fix these. Then I'll re-open it.

CoMPaTech commented:
Not sure how long you'd keep it close, but you can also reopen and convert to draft, otherwise the doc-PR might become unlinked too.

LaggAt committed:
cleanups for review

LaggAt committed:
Merge commit '6102d053066c5c439acfa06bd1a71c67a397a82c' into feature/govee_learn_brightness

LaggAt committed:
Merge branch 'feature/govee_learn_brightness' into feature/govee-led-strips

LaggAt committed:
manage online state, cleanups

LaggAt committed:
Merge commit 'f2b73c05d6884597d226397916b74d9189cfb0ab' into feature/govee-led-strips

LaggAt committed:
black fix

LaggAt committed:
pylint fix

LaggAt committed:
Merge remote-tracking branch 'upstream/dev' into feature/govee-led-strips

LaggAt commented:
_docs-missing_ isn't removed automatically (was also added manually, so this makes sense). 
Anything else shows green now.
Would be great to see it accepted :) Thanks.

CoMPaTech reviewed (commented):
You should double-check, although the lights show up green `config_flow` as indicated previously and automatically commented on the 'files changed' tab is not 100%. Unfortunately the 'CI/Process test coverage' doesn't pick it up, but it's there if you sift through.

```
homeassistant/components/govee/config_flow.py                             35      1    97%
```

There is no test for the last line of the code.

Again just quickly glancing at your code:

 - is `light` the only platform or are you preparing for others (like sensor) as well?
 - while you could import into const and use from there, I think HA mostly directly includes (but other reviewer might comment on that)
 - the `_async_update` in `GoveeDataUpdateCoordinator` looks a bit complicated, but if I recall correctly unless there is a very specific reason you should use your modules exceptions first and not just catch any exception.

LaggAt commented:
Thank your @CoMPaTech , 
i converted it to draft and will look into these. 

I didn't see that coverage locally, can I test this in the dev container somehow or do I need to check this manually?

LaggAt committed:
review fixes @CoMPaTech

LaggAt committed:
Merge commit 'd35e33610a8ca62af3b0c343403d5226ed870d7c' into feature/govee-led-strips

LaggAt commented:
> There is no test for the last line of the code.

right, I eliminated that line as unnecessary. 

> * is `light` the only platform or are you preparing for others (like sensor) as well?

yes, this is the only platform that will be implemented for now. I do not own other Govee devices.

> * while you could import into const and use from there, I think HA mostly directly includes (but other reviewer might comment on that)

yes, that's quick and ugly. I changed that.

> * the `_async_update` in `GoveeDataUpdateCoordinator` looks a bit complicated, but if I recall correctly unless there is a very specific reason you should use your modules exceptions first and not just catch any exception.

right, now I'm catching a GoveeError there, I just added this exception type and have overseen this code.
I also did add an event to the library, so that I can handle online/offline state change more easily and clear. So some logic could be removed there.

LaggAt committed:
Merge commit '1e9e40bf71720384623e644b7c4fce0181650109' into feature/govee-led-strips

CoMPaTech commented:
Keeping it open with the only lights in platform is a good thing, especially if the brand offers more platform based devices in the near future. Other (potential) devs can hook in than :)

I usually quicktest with this line from the main repo-dir: `pytest -rP --cov=homeassistant/components/YOURINTEGRATION/ --cov-report term-missing -- tests/components/YOURINTEGRATION/` to have all test-data and a summary of which lines in coverage ~~I'd miss out~~ I should improve on.

LaggAt committed:
linter fixes, get states immediatly when online

LaggAt committed:
github actions errored, pushing small change

LaggAt committed:
github actions errored, pushing small change

LaggAt committed:
Merge commit '42fb0e95457a2aee2911061a52424b685f792f14' into feature/govee-led-strips

LaggAt commented:
github actions seem to fail for some reason today. I am merging again with latest upstream/dev, if they fail again I'll look at it tomorrow. Thanks so far your all your input.

LaggAt committed:
config flow coverage 100%

LaggAt committed:
Merge commit '343e5d64b82b61a96051b3bf8994037341489b63' into feature/govee-led-strips

LaggAt committed:
removed unneeded import in test class

CoMPaTech reviewed (commented):
Looking good

LaggAt commented:
Fine. Is there anything else to do for me?

CoMPaTech commented:
You'll have to wait for some of the core team to have time to look at it, not sure how to move it to 'needs review' from incoming, I think some of the devs walk them and move them every now and again

LaggAt committed:
Merge branch 'dev' into feature/govee-led-strips

LaggAt committed:
Merge commit 'e0df91ef' into feature/govee-led-strips

LaggAt committed:
Merge branch 'feature/govee-led-strips' of https://github.com/LaggAt/home-assistant-core into feature/govee-led-strips

disforw commented:
Excited!

LaggAt committed:
bumped library version to 0.1.28

LaggAt committed:
Merge remote-tracking branch 'upstream/dev' into feature/govee-led-strips

LaggAt commented:
requirements_test_all.txt conflict solved,
bumped library version to 0.1.28

LaggAt committed:
fixed order in requirements

LaggAt committed:
Update README.rst

TheJefe commented:
It's not clear what is left todo to get this merged. Anything I can do to help?

LaggAt commented:
I've set this to draft, and paused my effort here.
For now I started a HACS repository where I could publish features and bug fixes.

ntindle commented:
Is there anything that needs to be done to pull this out of draft state?

janiversen commented:
Is this PR going to continue as “draft”, or will it be marked as “ready for review”.

In case it is going to change status, then it would be better to close it.

dominik-ba commented:
I'm sorry if this is a stupid newbie question, but wouldn't it be desirable to get the interpretation into core and have hacs only if that for dinner reason does not work?
